man/man1/ring.1
.TH RING 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
ring/ringd \- message client & server for W
.SH SYNOPSIS
.nf
ringd &
ring [<server address>] <message>
.fi
.SH DESCRIPTION
\fIRing\fP will call the \fIringd\fP server at the given address and
sends it the message. Ringd server will show all the messages it receives
in the windows it pops up into the middle of the screen or on standard
error if it's unable to connect W server.
.PP
You can use this for sending and receiving little notices like "I'm
going for snack, meet me at the front door" between friends.  It's an
alternative for e-mail and \fIwbiff\fP.  ;-)
.SH OPTIONS
If there are more than one argument, \fIring\fP will use the first one
as the address of machine where to send the message.  'localhost' is the
default address.  Message is rest of the arguments.  It's better to
always quote what you're sending so that shell won't accidentally expand
$ and other special characters in your message.
.SH AUTHOR
(w) 11/1996 by Eero Tamminen
.SH SEE ALSO
.IR wbiff (1)
.PP
Check out \fIUsing C on the UNIX system\fP by \fBO'Reilly\fP if you're
interested where I got the socket code.  These programs were the first
socket programs I have made.



wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/ring.c
/*
 * ring.c, a part of the W Window System
 *
 * Copyright (C) 1996 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a 'ring' client (actually a minimal socket example)
 *
 * sends a msg over network to given ring server which then
 * pops up a window with the message.
 *
 * Ref: "Using C on the UNIX system", O'Reilly & Assiciates, Inc.
 */

#include <stdio.h>
#include <unistd.h>
#ifdef LOCAL_SERVER
#include <sys/un.h>		/* AF_UNIX */
#else
#include <netinet/in.h>		/* AF_INET */
#endif
#include <sys/time.h>		/* fd_set */
#include <sys/socket.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <netdb.h>
#include <string.h>

#define RING_SOCK	"/tmp/ringd"
#define RING_PORT	21060		/* "RD" like in Ring Daemon */

int main(int argc, char *argv[])
{
  int sock;
  short lenght;
  const char *string;

#ifdef LOCAL_SERVER

  struct sockaddr_un addr;
  int len;

  if(argc != 2)
  {
    fprintf(stderr, "\nusage: %s <message>\n", *argv);
    fprintf(stderr, "sends message to the 'ring' daemon\n");
    return -1;
  }

  if((sock = socket(AF_UNIX, SOCK_STREAM, 0)) < 0)
  {
    fprintf(stderr, "can't create AF_UNIX socket.\n");
    return -1;
  }

  addr.sun_family = AF_UNIX;
  strcpy(addr.sun_path, RING_SOCK);
  len = sizeof(addr.sun_family) + strlen(addr.sun_path);
  if(connect(sock, (struct sockaddr *)&addr, len))
  {
    fprintf(stderr, "can't connect to local " RING_SOCK " socket.\n");
    close(sock);
    return -1;
  }
  string = argv[1];

#else

  struct hostent *host;
  struct sockaddr_in addr;

  if(argc < 2 || argc > 3 || argv[1][0] == '-')
  {
    fprintf(stderr, "\nusage: %s [address] <message>\n", *argv);
    fprintf(stderr, "sends message to the 'ring' daemon at given address\n");
    return -1;
  }

  if(argc == 3)
    string = argv[1];
  else
    string = "localhost";

  /* resolv host address */
  if(!(host = gethostbyname(string)))
  {
    fprintf(stderr, "can't resolve adress `%s'.\n", string);
    return -1;
  }

  /* create non-local stream socket */
  if((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0)
  {
    fprintf(stderr, "can't create AF_INET socket.\n");
    return -1;
  }

  /* connect specified address */
  addr.sin_family = AF_INET;
  addr.sin_port = htons(RING_PORT);
  addr.sin_addr = *(struct in_addr *)host->h_addr_list[0];
  if(connect(sock, (struct sockaddr *)&addr, sizeof(struct sockaddr)))
  {
    fprintf(stderr, "can't connect to server `%s' (%s).\n",
      string, inet_ntoa(addr.sin_addr));
    close(sock);
    return -1;
  }

  if(argc == 3)
    string = argv[2];
  else
    string = argv[1];

#endif

  /* send message with it's lenght */
  lenght = htons(strlen(string));
  write(sock, &lenght, sizeof(lenght));
  write(sock, string, strlen(string));
  close(sock);

  return 0;
}
